builder/chroots.go

```diff
@@ -0,0 +1,637 @@
+
+	err = os.MkdirAll(chrootVersionDir, 0755)
+	if err != nil {
+		return err
+	}
+	for name, bundle := range set.Bundles {
+		// TODO: Should we embed this information in groups.ini? (Maybe rename it to bundles.ini)
+		var includes bytes.Buffer
+		for _, inc := range bundle.DirectIncludes {
+			fmt.Fprintf(&includes, "%s\n", inc)
+		}
+		err = ioutil.WriteFile(filepath.Join(chrootVersionDir, name+"-includes"), includes.Bytes(), 0644)
+		if err != nil {
+			return err
+		}
+	}
+
+	fmt.Println("Creating os-core chroot")
+	osCoreDir := filepath.Join(chrootVersionDir, "os-core")
+	err = os.MkdirAll(filepath.Join(osCoreDir, "var/lib/rpm"), 0755)
+	if err != nil {
+		return err
+	}
+
+	fmt.Println("Initializing RPM database")
+	_, err = run(
+		"rpm",
+		"--root", osCoreDir,
+		"--initdb",
+	)
+	if err != nil {
+		return err
+	}
+
+	yumCmd := []string{
+		"yum",
+		"--config=" + b.Yumconf,
+		"-y",
+		"--releasever=" + b.Clearver,
+	}
+	// When in Fedora, call dnf instead of yum.
+	if osInfo, oerr := readOSInfo(); oerr == nil {
+		if osInfo.ID == "fedora" {
+			// TODO: Simplify this when we can assume all Fedora users will be >= 22?
+			versionID, _ := strconv.Atoi(osInfo.VersionID)
+			if versionID >= 22 {
+				yumCmd[0] = "dnf"
+			}
+		}
+	}
+
+	fmt.Println("Installing filesystem package in os-core")
+	_, err = run(merge(yumCmd,
+		"--installroot="+osCoreDir,
+		"install",
+		"filesystem",
+	)...)
+	if err != nil {
+		return err
+	}
+
+	clearDir := filepath.Join(osCoreDir, "usr/share/clear")
+	err = os.MkdirAll(filepath.Join(clearDir, "bundles"), 0755)
+	if err != nil {
+		return err
+	}
+
+	fmt.Println("Installing packages in os-core")
+	err = installPackagesToBundleChroot(yumCmd, chrootVersionDir, set.Bundles["os-core"])
+	if err != nil {
+		return err
+	}
+
+	// Writing special files identifying the version in os-core.
+	err = ioutil.WriteFile(filepath.Join(clearDir, "version"), []byte(version), 0644)
+	if err != nil {
+		return err
+	}
+	// TODO: This seems to be the only thing that makes two consecutive chroots of the same
+	// version to be different. Use SOURCE_DATE_EPOCH if available?
+	versionstamp := fmt.Sprint(time.Now().Unix())
+	err = ioutil.WriteFile(filepath.Join(clearDir, "versionstamp"), []byte(versionstamp), 0644)
+	if err != nil {
+		return err
+	}
+
+	fmt.Println("Creating 'versions' file")
+	err = createVersionsFile(chrootVersionDir, yumCmd)
+	if err != nil {
+		return errors.Wrapf(err, "couldn't create the versions file")
+	}
+
+	err = fixOSRelease(filepath.Join(osCoreDir, "usr/lib/os-release"), version)
+	if err != nil {
+		return errors.Wrap(err, "couldn't fix os-release file")
+	}
+
+	fmt.Println("Creating chroots for bundles")
+	// TODO: Use goroutines.
+	for _, bundle := range set.Bundles {
+		if bundle.Name == "os-core" {
+			continue
+		}
+
+		fmt.Printf("Creating %s chroot\n", bundle.Name)
+		_, err = run("cp", "-a", "--preserve=all", osCoreDir, filepath.Join(chrootVersionDir, bundle.Name))
```

Comment:
Is there not a way to do this without shelling out to `cp`?

Reply:
quick research shows you would have to implement your own recursive copy... never mind.